pypesto/objective/amici_prediction.py

```diff
@@ -0,0 +1,439 @@
+    requested. This responsibility for correctness of this sensitivity
+    post-processing is EXPLICITLY left to the user.
+    """
+    def __init__(self,
+                 amici_objective: AmiciObjective,
+                 post_processing: Union[Callable, None] = None,
+                 post_processing_sensi: Union[Callable, None] = None,
+                 post_processing_timepoints: Union[Callable, None] = None,
+                 max_num_conditions: int = 0,
+                 observables: Sequence[str] = None):
+        """
+        Constructor.
+
+        Parameters
+        ----------
+        amici_objective:
+            An objective object, which will be used to get model simulations
+        post_processing:
+            A callable function which applies postprocessing to the simulation
+            results. Default are the observables of the amici model.
+            This method takes a list of ndarrays (as returned in the field
+            ['y'] of amici ReturnData objects) as input.
+        post_processing_sensi:
+            A callable function which applies postprocessing to the
+            sensitivities of the simulation results. Default are the
+            observable sensitivities of the amici model.
+            This method takes two lists of ndarrays (as returned in the
+            fields ['y'] and ['sy'] of amici ReturnData objects) as input.
+        post_processing_timepoints:
+            A callable function which applies postprocessing to the timepoints
+            of the simulations. Default are the timepoints of the amici model.
+            This method takes a list of ndarrays (as returned in the field
+            ['t'] of amici ReturnData objects) as input.
+        max_num_conditions:
+            In some cases, we don't want to compute all predictions at once
+            when calling the prediction function, as this might not fit into
+            the memory for large datasets and models.
+            Here, the user can specify a maximum number of conditions, which
+            should be simulated at a time.
+            Default is 0 meaning that all conditions will be simulated.
+            Other values are only applicable, if an output file is specified.
+        observables:
+            IDs of observables, if post-processing is used
+        """
+        # save settings and objective
+        self.amici_objective = amici_objective
+        self.max_num_conditions = max_num_conditions
+        self.post_processing = post_processing
+        self.post_processing_sensi = post_processing_sensi
+        self.post_processing_timepoints = post_processing_timepoints
+
+        if observables is None:
+            self.observables = amici_objective.amici_model.getObservableIds()
+        else:
+            self.observables = observables
+
+    def __call__(
+            self,
+            x: np.ndarray,
+            sensi_orders: Tuple[int, ...] = (0, ),
```

Comment:
Although, maybe this doesn't apply to (immutable) tuples :thinking:

Reply:
Hm... thought about it, but I'd like to keep that defaulted in the same way as in the AmiciObjective call... In this case, `(0,)` does not correspond to something like `None`, so this would also be not an intuitive default.
Moreover, the sensi_order should always be read-only. There should never be a change of it, so I think it's fine to keep it. But the problem that is pointed out in the link would in principle also apply to tuples. Being immutable just means that their elements can't be changed. Concatenation or things like that are still possible.

Reply:
Just did a quick check and it looks like it would only be an issue for tuples with mutable elements, like `([],)`, as the list object within the tuple can have elements modified with each call (or by modifying the return object). But in the case of `(0,)`, this doesn't appear to be an issue. So, fine for me :+1: